waiter/src/waiter/process_request.clj

```diff
@@ -252,13 +252,26 @@
                                          (let [complete-trigger-id (utils/unique-identifier)]
                                            (deliver complete-triggered-promise complete-trigger-id)
                                            (when (= complete-trigger-id @complete-triggered-promise)
-                                             (log/debug "closing request body as" message)
+                                             ;; avoid decrementing the counter multiple times
                                              (counters/dec! request-body-streaming-counter)
-                                             (when throwable
-                                               (log/error throwable "unable to stream request bytes, aborting request")
-                                               (async/>!! abort-ch throwable))
-                                             (report-request-size-metrics 0 true)
-                                             (async/close! body-ch))))))]
+                                             (report-request-size-metrics 0 true))
+                                           (if throwable
+                                             (let [identifier-map {:identifier complete-trigger-id}
+                                                   ;; the callback is necessary as there is a data race between aborting the
+                                                   ;; request and closing of the body channel triggering a normal complete before
+                                                   ;; the abort request gets processed.
+                                                   callback (fn complete-request-streaming-abort-result-callback [aborted?]
+                                                              (cid/with-correlation-id
+                                                                correlation-id
+                                                                (log/info "result of aborted request:" aborted? identifier-map)
+                                                                (async/close! body-ch)))]
+                                               (log/info throwable "aborting request" message identifier-map)
+                                               (when-not (async/put! abort-ch [throwable callback])
+                                                 ;; abort channel already closed, cleanup by closing the body channel
+                                                 (async/close! body-ch)))
```

Comment:
Is it still possible that we're closing the `body-ch` before aborting the request?

Reply:
Can you point me to the code that handles aborting the response to the client such that on the close of `body-ch` we don't complete normally?

Reply:
I'm not sure I understand the question completely. Closing `body-ch` means we are done streaming the request body. It does not directly trigger the abort of the backend request.

Invoking abort on the backend request causes Jetty to treat the request as a failure and stream appropriate frames to the client in the response body. The abort is called at:
https://github.com/twosigma/jet/blob/e1f0a7b279c564c4a457ee1d020797c818bc7f39/src/clj/qbits/jet/client/http.clj#L434

and the backend request cleanup is performed at:
https://github.com/twosigma/jet/blob/e1f0a7b279c564c4a457ee1d020797c818bc7f39/src/clj/qbits/jet/client/http.clj#L547

The above causes the exception to be propagated on the `error-chan` which we process at:
https://github.com/twosigma/waiter/blob/44685bb6d22fc242f9a741f25f5e7c12c00e330e/waiter/src/waiter/process_request.clj#L446